zcash_client_backend/src/data_api/ll/wallet.rs

```diff
@@ -112,13 +118,73 @@ where
     })
 }
 
+#[cfg(feature = "transparent-inputs")]
+pub enum GapAddressesError<SE> {
+    Storage(SE),
+    AddressGeneration(AddressGenerationError),
+    AccountUnknown,
+}
+
+#[cfg(feature = "transparent-inputs")]
+pub fn generate_transparent_gap_addresses<DbT, SE>(
+    wallet_db: &mut DbT,
+    gap_limits: GapLimits,
+    account_id: <DbT as LowLevelWalletRead>::AccountId,
+    key_scope: TransparentKeyScope,
+    request: UnifiedAddressRequest,
+) -> Result<(), GapAddressesError<SE>>
+where
+    DbT: LowLevelWalletWrite<Error = SE>
+        + GapLimitsWalletAccess<Error = SE, AccountRef = <DbT as LowLevelWalletRead>::AccountRef>,
+    DbT::TxRef: Eq + Hash,
+{
+    let account_ref = wallet_db
+        .get_account_ref(account_id)
+        .map_err(GapAddressesError::Storage)?;
+
+    let gap_limit = match key_scope {
+        TransparentKeyScope::EXTERNAL => Ok(gap_limits.external()),
+        TransparentKeyScope::INTERNAL => Ok(gap_limits.internal()),
+        TransparentKeyScope::EPHEMERAL => Ok(gap_limits.ephemeral()),
+        _ => Err(AddressGenerationError::UnsupportedTransparentKeyScope(
+            key_scope,
+        )),
+    }
+    .map_err(GapAddressesError::AddressGeneration)?;
+
+    if let Some(gap_start) = wallet_db
+        .find_gap_start(account_ref, key_scope, gap_limit)
+        .map_err(GapAddressesError::Storage)?
+    {
+        let account = wallet_db
+            .get_account_internal(account_ref)
+            .map_err(GapAddressesError::Storage)?
+            .ok_or(GapAddressesError::AccountUnknown)?;
+        wallet_db
+            .generate_address_range_internal(
+                account_ref,
+                &account.uivk(),
+                account.ufvk(),
+                key_scope,
+                request,
+                gap_start..gap_start.saturating_add(gap_limit),
+                false,
+            )
```

Comment:
One of the primary things I would like to see is for all the "business logic" parts of `generate_address_range_internal` to be moved up into `zcash_keys`, which AFAICT this PR does not yet do. The implementation here, which involves a few IO operations against the backing data store that should be abstracted over, is what needs to be "pulled up" into `zcash_keys` and made reusable. What's currently being moved up is kind of trivial.

Reply:
Thanks for doing this work, I will re-review it tomorrow!